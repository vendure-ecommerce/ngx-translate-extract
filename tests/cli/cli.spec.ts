import { exec } from 'node:child_process';
import { readdir, readFile, rm } from 'node:fs/promises';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';

import { afterAll, beforeAll, describe, test } from 'vitest';

const execAsync = promisify(exec);
const __dirname = dirname(fileURLToPath(import.meta.url));

const REPOSITORY_ROOT = resolve(__dirname, '../..');
const FIXTURES_PATH = resolve(REPOSITORY_ROOT, 'tests/cli/fixtures/');
const TMP_PATH = resolve(REPOSITORY_ROOT, 'tests/cli/tmp');
const CLI_PATH = resolve(TMP_PATH, 'dist/cli/cli.js');

let nextFileId = 0;
const createUniqueFileName = (fileName: string) => resolve(TMP_PATH, `${nextFileId++}-${fileName}`);

describe.concurrent('CLI Integration Tests', () => {
	beforeAll(async () => {
		try {
			await execAsync(`npm run build -- --outDir ${TMP_PATH}/dist`);
		} catch (err) {
			console.error('Error during build in beforeAll:', err);
			throw err;
		}
	});

	afterAll(async () => {
		await rm(TMP_PATH, { recursive: true });
	});

	test('shows the expected output when extracting', async ({expect}) => {
		const OUTPUT_FILE = createUniqueFileName('strings.json');
		const fixtureFiles = await readdir(FIXTURES_PATH);
		const { stdout } = await execAsync(`node ${CLI_PATH} --input ${FIXTURES_PATH} --output ${OUTPUT_FILE} --format=json`);

		expect(stdout).toContain('Extracting:');
		fixtureFiles.forEach(file => expect(stdout).toContain(file));

		expect(stdout).toContain('Found 15 strings.');
		expect(stdout).toContain('Saving:');

		expect(stdout).toContain(OUTPUT_FILE);
		expect(stdout).toContain('Done.');
	})

	test('extracts translation keys to a .json file', async ({ expect }) => {
		const OUTPUT_FILE = createUniqueFileName('strings.json');
		await execAsync(`node ${CLI_PATH} --input ${FIXTURES_PATH} --output ${OUTPUT_FILE} --format=json`);

		const extracted = await readFile(OUTPUT_FILE, { encoding: 'utf8' });

		expect(extracted).toMatchSnapshot();
	});

	test('extracts translation keys to a .po file', async ({ expect }) => {
		const OUTPUT_FILE = createUniqueFileName('strings.po');
		await execAsync(`node ${CLI_PATH} --input ${FIXTURES_PATH} --output ${OUTPUT_FILE} --format=pot`);

		const extracted = await readFile(OUTPUT_FILE, { encoding: 'utf8' });

		expect(extracted).toMatchSnapshot();
	});

	test('extracts translation keys to a .po file without file location comments', async ({ expect }) => {
		const OUTPUT_FILE = createUniqueFileName('strings.po');
		await execAsync(`node ${CLI_PATH} --input ${FIXTURES_PATH} --output ${OUTPUT_FILE} --format=pot --no-po-source-locations`);

		const extracted = await readFile(OUTPUT_FILE, { encoding: 'utf8' });

		expect(extracted).toMatchSnapshot();
	});
});
