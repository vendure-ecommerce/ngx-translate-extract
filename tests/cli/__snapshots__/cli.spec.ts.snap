// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI Integration Tests > extracts translation keys to a .json file 1`] = `
"{
	"translate-service.comp.welcome": "",
	"translate-service.comp.description": "",
	"translate-service.comp.details": "",
	"translate-service.comp.show-more-label": "",
	"translate-service.comp.show-less-label": "",
	"pipe.comp.welcome": "",
	"pipe.comp.description": "",
	"directive.comp.welcome": "",
	"directive.comp.description": "",
	"private-translate-service.comp.welcome": "",
	"private-translate-service.comp.description": "",
	"ngx-translate.marker.dashboard.title": "",
	"ngx-translate.marker.dashboard.description": "",
	"marker.dashboard.title": "",
	"marker.dashboard.description": ""
}"
`;

exports[`CLI Integration Tests > extracts translation keys to a .po file 1`] = `
"msgid ""
msgstr ""
"mime-version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"

#: ngx-translate-extract/tests/cli/fixtures/translate-service.component.fixture.ts
msgid "translate-service.comp.welcome"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/translate-service.component.fixture.ts
msgid "translate-service.comp.description"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/translate-service.component.fixture.ts
msgid "translate-service.comp.details"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/translate-service.component.fixture.ts
msgid "translate-service.comp.show-more-label"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/translate-service.component.fixture.ts
msgid "translate-service.comp.show-less-label"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/translate-pipe.component.fixture.ts
msgid "pipe.comp.welcome"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/translate-pipe.component.fixture.ts
msgid "pipe.comp.description"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/translate-directive.component.fixture.ts
msgid "directive.comp.welcome"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/translate-directive.component.fixture.ts
msgid "directive.comp.description"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/private-translate-service.component.fixture.ts
msgid "private-translate-service.comp.welcome"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/private-translate-service.component.fixture.ts
msgid "private-translate-service.comp.description"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/ngx-translate.marker.fixture.ts
msgid "ngx-translate.marker.dashboard.title"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/ngx-translate.marker.fixture.ts
msgid "ngx-translate.marker.dashboard.description"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/marker.fixture.ts
msgid "marker.dashboard.title"
msgstr ""

#: ngx-translate-extract/tests/cli/fixtures/marker.fixture.ts
msgid "marker.dashboard.description"
msgstr """
`;

exports[`CLI Integration Tests > extracts translation keys to a .po file without file location comments 1`] = `
"msgid ""
msgstr ""
"mime-version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"

msgid "translate-service.comp.welcome"
msgstr ""

msgid "translate-service.comp.description"
msgstr ""

msgid "translate-service.comp.details"
msgstr ""

msgid "translate-service.comp.show-more-label"
msgstr ""

msgid "translate-service.comp.show-less-label"
msgstr ""

msgid "pipe.comp.welcome"
msgstr ""

msgid "pipe.comp.description"
msgstr ""

msgid "directive.comp.welcome"
msgstr ""

msgid "directive.comp.description"
msgstr ""

msgid "private-translate-service.comp.welcome"
msgstr ""

msgid "private-translate-service.comp.description"
msgstr ""

msgid "ngx-translate.marker.dashboard.title"
msgstr ""

msgid "ngx-translate.marker.dashboard.description"
msgstr ""

msgid "marker.dashboard.title"
msgstr ""

msgid "marker.dashboard.description"
msgstr """
`;
